server/src/core/search/search_executor/preprocess.rs

```diff
@@ -0,0 +1,297 @@
+#[derive(Debug, Clone)]
+struct SearchFilter {
+    parent: Option<Vec<String>>,
+    r#type: Option<Vec<String>>,
+    usage: Option<Vec<String>>,
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct SearchInput {
+    pub(super) tokens: Vec<SearchToken>,
+    #[allow(dead_code)]
+    filter: SearchFilter,
+}
+
+impl SearchInput {
+    pub fn to_query_string(self) -> String {
+        let mut s = String::from("");
+        for token in self.tokens.iter() {
+            if token.closed && !token.quoted {
+                s.push_str(&format!("{} ", token.s));
+            } else {
+                s.push_str(&token.s.clone());
+            }
+        }
+
+        s
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub(super) struct InputToken {
+    s: String,
+    regular_split: bool,
+    closed: bool,
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct SearchToken {
+    pub(super) s: String,
+    #[allow(dead_code)]
+    regular_split: bool,
+    pub(super) closed: bool,
```

Comment:
Yes, only the last token can have `closed` false

Reply:
Should I refactor this name to `last_token` or remove it and check this property in the loops generating the joined querry strings?

Reply:
Hm, maybe there is a better way to name this, but the semantic meaning of `closed` is different to `last_token`. The reason I added it there is because MeiliSearch makes a difference between a space at the end or not.

Giving '`mathe `' (with a space at the end) to MeiliSearch will not find the MI, because MeiliSearch only does 'prefix search' on the last token and only if there is not space at the end. However if we just always dropped any spaces at the end, I found it irritating that it still does prefix search after hitting space (see how Google does change autosuggestions as well when you hit space).

The idea of this property is to tell the function that generates the strings for MeiliSearch whether that token has been 'closed' with a space.

For quoted tokens the meaning is different. The following queries have closed=false for the last token:
'`"my`', '`"my `', '`"my query`',
and the following closed=true:
'`"my query"`', '`"my query" `'